### PR TITLE
🅿️®️🅾️❗️

### LINKED ISSUE #18: Import / Export JSON file of blocks
This is especially useful for theme authors who wish to use ACB to build blocks, and want to distribute the blocks with their themes.

### PR DESCRIPTION
**Pull request by lukecarbis**
This PR introduces:

- [x] A Settings screen with offsite links to Docs and Help
- [x] Licensing for Pro registration
- [x] Export individual blocks
- [x] Bulk export blocks
- [x] Import blocks (via Tools > Import) (note: we can improve visibility of this feature in a future release)
- [x] A "Go Pro" feature (visible when there is no active license) (To be completed by @RobStino)

@kienstra I have a question about how the plugin is structured in this PR. I've stored additional classes (`class-admin.php`, `class-import.php`, `class-settings.php`, `class-settings.php` and `class-license.php`) in the `php` directory, but I'd like your opinion on whether there should be a (namespaced?) child directory within `php` where these files are kept.

Similarly, are you happy with `php/views/` for front-endy template files?

Closes #18.

### CODE REVIEW

File: php/class-settings.php
Comment (kienstra) [inline php/class-settings.php]: Should this have the textdomain?
```php
<?php esc_html_e( 'License', 'block-lab' ); ?>
```
And likewise for the `esc_html_e()` calls below.

File: php/class-upgrade.php
Comment (kienstra) [inline php/class-upgrade.php]: This PR is marked WIP, so maybe this method will later have logic.

File: php/post-types/class-block-post.php
Comment (kienstra) [inline php/post-types/class-block-post.php:1026]: It's nice how this uses `sprintf()` to construct the `<a ...>`.

File: php/views/license.php
Comment (kienstra) [inline php/views/license.php]: This is fine as is. But it might be simpler to include the `<strong>` in the `__()` call:
```php
__( 'Your license includes <strong>%s</strong> site installs.', 'block-lab' )
```

This [happens in Core](https://github.com/WordPress/wordpress-develop/blob/9448313544e5d6db16ac640d552258405e8bc3d8/src/wp-admin/revision.php#L128) sometimes.

Likewise for the similar `__()` calls below.

File: php/views/upgrade.php
Comment (kienstra) [inline php/views/upgrade.php:1]: Nice organization of this in `views/`

File: php/class-upgrade.php
Comment (kienstra) [inline php/class-upgrade.php]: This isn't my domain 😄 

But do you think some people might confuse this with the [GoPro](https://gopro.com) camera?

go-pro

File: php/views/license.php
Comment (kienstra) [inline php/views/license.php]: Wrapping `add_query_arg()` in `esc_url()` might be a little safer.

File: php/class-license.php
Comment (kienstra) [inline php/class-license.php]: This works well when entering a wrong license key:
 
feedback-license

File: php/class-import.php
Comment (kienstra) [inline php/class-import.php]: The import UI for the pro version looks good, though I didn't test it with a JSON file:


import-ui

File: css/admin.css
Comment (kienstra) [inline css/admin.css:1]: It's nice how you organized the stying in `css/`. Sometimes I'll output inline `<style>` tags in the markup, but it's cleaner how they're organized here in `css/`.

File: php/class-import.php
Comment (kienstra) [inline php/class-import.php]: Maybe this function will be populated later.

File: php/class-license.php
Comment (kienstra) [inline php/class-license.php]: How about `DAY_IN_SECONDS` here?

It looks like [it's defined in Core](https://codex.wordpress.org/Easier_Expression_of_Time_Constants) (I just found out about this).

File: php/class-settings.php
Comment (kienstra) [inline php/class-settings.php]: This is an interesting way of queuing the notices to display 😄

**Review by kienstra — COMMENTED** [review]
**Looks Good**
*Some Minor Points*

Hi @lukecarbis,
This PR for the pro license looks really good. 

It's WIP and there are some small points, so I won't approve it yet if that's alright.

>@kienstra I have a question about how the plugin is structured in this PR. I've stored additional classes (`class-admin.php`, `class-import.php`, `class-settings.php`, `class-settings.php` and `class-license.php`) in the `php` directory, but I'd like your opinion on whether there should be a (namespaced?) child directory within `php` where these files are kept.

How about a `php/admin/` directory for those files? The AMP plugin has something similar in [includes/admin/](https://github.com/ampproject/amp-wp/tree/286aef5d61e838d4b4b5797acd583a5e0e947f90/includes/admin). 

Of course, many more files relate to `/wp-admin`, but these seem specific to admin menus.

Like you mentioned, it'd be good to namespace the files if they go in a directory like `php/admin/`, to follow the convention of the `php/blocks/` files having a namespace of `Block_Lab\Blocks`, etc...

So maybe the namespace could be something like `Block_Lab\Admin`.

>Similarly, are you happy with `php/views/` for front-endy template files?

Yes, good idea to put them in `php/views/`. If you go with something like a `php/admin/` directory, maybe they could be in `php/admin/views/`. But I think they're fine either way.

File: php/post-types/class-block-post.php
Comment (lukecarbis) [reply php/post-types/class-block-post.php:1026]: Thanks!

File: php/views/license.php
Comment (lukecarbis) [reply php/views/license.php]: Great pickup.

File: php/class-settings.php
Comment (lukecarbis) [reply php/class-settings.php]: Whoops! Thanks for noticing.

File: php/class-license.php
Comment (lukecarbis) [reply php/class-license.php]: Oh I totally forgot about that! Yes, definitely.

File: php/class-settings.php
Comment (lukecarbis) [reply php/class-settings.php]: Yes, this happens this way for a reason. I am using WordPress' Settings API to save all the options, which means I don't have to write any code which checks for $_POST variables or any code to `update_option` s.

The side effect is that the form goes off to `options.php` to save, and then redirects back to the settings page. After the redirect, I've got no $_POST variables to speak of, so I don't know  if the form was saved, or if it was saved successfully.

Two ways I see around this:
- Hijack the options saving process and redirect back to the settings page with an additional $_GET parameter indicating the result (then using that result to display a notice)
- Save the notice in the database, and check if it exists when the page loads (and show if it does)

I opted for the second method, since I feel it's cleaner, and potentially reusable in other situations.

**Comment by lukecarbis** [comment]
@kienstra Ready for you to re-review. No longer a [WIP]!

I've moved Admin, License, Settings, and Upgrade to `php/admin/` and given them the `Admin` namespace.

Only problem now is that we have a class called `Admin\Admin`, which is a bit repetitive. 

Can you think of a better name for the `Admin\Admin` class, or the namespace we're using? Or do you think this classname is okay?

File: php/class-settings.php
Comment (kienstra) [reply php/class-settings.php]: Sounds good, thanks for explaining this 😄

**Comment by kienstra** [comment]
>Can you think of a better name for the `Admin\Admin` class, or the namespace we're using? Or do you think this classname is okay?

Maybe `Admin\Setup`.

Even `Admin\Admin` might be an alright name, though. There's a [Blocks\Block](https://github.com/getblocklab/block-lab/blob/8c8d31d260ca143ea91970f2c207c741d69d1fe3/php/blocks/class-block.php#L15) class,

**Comment by kienstra** [comment]
>I've moved Admin, License, Settings, and Upgrade to `php/admin/` and given them the `Admin` namespace.

Nice 😄

**Comment by kienstra** [comment]
**Pro Page**

@lukecarbis,
It looks like the error message isn't displaying correctly on submitting a wrong license key:

error-message

**Comment by lukecarbis** [comment]
@kienstra That should be fixed now.

**Comment by kienstra** [comment]
**Approved**

Hi @lukecarbis,
This looks really good. Thanks for addressing the points here.

The notice looks good, too:
notice-appears-expected

**Review by kienstra — APPROVED** [review]

**Pull request merged**
